Realm/Realm/DatabaseTypes/RealmCollectionBase.cs

```diff
@@ -204,16 +204,30 @@ protected void AddToRealmIfNecessary(in RealmValue value)
                 return;
             }
 
-            var robj = value.AsRealmObject<IRealmObject>();
+            var robj = value.AsRealmObject<IRealmObjectBase>();
 
             if (robj.IsManaged && !robj.Realm.IsSameInstance(Realm))
             {
                 throw new RealmException("Can't add to the collection an object that is already in another realm.");
             }
 
-            if (!robj.IsManaged)
+            switch (robj)
             {
-                Realm.Add(robj);
+                case IRealmObject topLevel:
+                    if (!robj.IsManaged)
+                    {
+                        Realm.Add(topLevel);
+                    }
+
+                    break;
+
+                // Embedded and asymmetric objects will not reach this path unless they are used as a RealmValue.
```

Comment:
I think what this comment is trying to say is that we should only hit this codepath if the property on the object is declared as `public IList<RealmValue> Foo { get; }`. While the argument here is `RealmValue`, this is because all types eventually convert to it, so you'll get here for `IList<int>` and `IList<Person>`.

Reply:
Given that no user code should arrive here, and it's on us to be sure about that, shouldn't we put some `Debug.Assert`?

Reply:
User code could arrive here, which is what the comment states - we'll only get here if a user is trying to add an embedded/asym object to a `Collection<RealmValue>`. We can add a debug assert to check the collection type is indeed `RealmValue`.

Reply:
Adding `Debug.Assert`s and clarifying the comment would be good :) I added the initial comment because it's not obvious that this code path only is reached in the event that the user uses embedded or asym. objects as a `RealmValue`. Since the argument itself is `RealmValue`, I see how it can get confused with what I was trying to say about "the user using it" :p 

Update suggestion:
```c#
// ...

// Embedded and asymmetric objects will not reach this path unless the user explicitly adds them
// to the collection as RealmValues (e.g. IList<RealmValue>).
case IEmbeddedObject:
    Debug.Assert(typeof(T) == typeof(RealmValue), "Expected a RealmValue to contain the IEmbeddedObject.");
    throw new NotSupportedException("A RealmValue cannot contain an embedded object.");
case IAsymmetricObject:
    Debug.Assert(typeof(T) == typeof(RealmValue), "Expected a RealmValue to contain the IAsymmetricObject.");
    throw new NotSupportedException("A RealmValue cannot contain an asymmetric object.");

// ...
```

Reply:
Yes yes, it was not obvious at all that there's a specific condition from the caller that drives what types come in. I'm not a fan of this design, but this is what we have. The comment is now clearer, thank you.